app/models/manager_refresh/inventory_collection.rb

```diff
@@ -146,6 +149,11 @@ def object_index_with_keys(keys, object)
       keys.map { |attribute| object.public_send(attribute).to_s }.join(stringify_joiner)
     end
 
+    def find_or_build(manager_uuid)
+      # FIXME: splat manager_ref
+      data_index[manager_uuid] || build({ manager_ref.first => manager_uuid })
```

Comment:
maybe the best would be to do find_or_build_by(:hardware => xy, :device_name => xx) or find_or_build_by(:ems_ref => xy) ?

with a check that only allowed combination is manager_ref. I will be bringing ability to create multiple indexes on the InventoryCollection later.